### ISSUE TITLE
Number of records returned is different from when manually querying Logs insight

### ISSUE THREAD

**Issue by lloyd-fftrf** [gogama/incite#16]
# User Story
When I use `incite` for querying CloudWatch Logs, the number of logs returned is different from when I use Logs Insight manual querying.

# Background
This is my `QuerySpec`:
```go
incite.QuerySpec{
	Text:       queryString,
        Start:      time.Date(year, month, day, hour, min, sec, nsec, location),
	End:        start.Add(5 * time.Minute),
	Groups:     []string{"aws/log/group"},
	Limit:      incite.MaxLimit,
	Chunk:      1 * time.Minute,
	SplitUntil: 1 * time.Minute,
}
```
After querying, I use `encoding/csv` package to format the result to `.csv`. This is my code snippet. Where `dataSlice` contains the slices of strings from the log query result.
```go
f, err := os.Create(fileName)
defer f.Close()
if err != nil {
	fmt.Println(err.Error())
	return
}

w := csv.NewWriter(f)
defer w.Flush()

for _, record := range dataSlice {
	err = w.Write(record)
	if err != nil {
		fmt.Println(err.Error())
		return
	}
}
```
When I used Logs Insight for query, it returns 13,006 records where if I use `incite` it would only return 12,767.
The result is consistent when I rerun the program.

**Comment by lloyd-fftrf** [comment]
Hi @vcschapp , sorry for the delayed reply, but this is the data that you requested

1. I was querying Apache logs  
  `fields @timestamp, @message | filter @logStream like /http_access.log_*/ | parse @message '*, * [*] \"*\" * * \"*\" \"*\"' as ip, f1, time, http_request, response, f2, url, user_agent | display ip, time, http_request, response, url, user_agent, f1, f2 | sort @timestamp asc`

2. `dataSlice` is obtained by reflecting the incite return values. I'm removing error catching for readability. The thing is, I already got `12767` when printing the `len(result)` (It shows `13006` when manually scanning the CloudWatch Logs Insight)
  ```go
  result, err := incite.ReadAll(stream)

  fmt.Printf("logs scanned: %v\n", len(result))

  var logStructSlice []logFormat
  err = incite.Unmarshal(result, &logStructSlice)

  var dataSlice [][]string
  isTitled := false

  // convert logStructSlice to dataSlice
  for _, logStruct := range logStructSlice {
  // create temporary slice to hold the values
  // of the dataSlice's elements
  var tempSlice []string

  // one time slice for inserting the title
  // (incite struct's key field)
  var titleSlice []string

  // change the element value to reflect.Value for processing
  rv := reflect.ValueOf(logStruct)

  // append each element's value to the temporary slice
  // titleSlice will be created only on the first iteration
  for i := 0; i < rv.NumField(); i++ {
    if !isTitled {
      columnTitle := rv.Type().Field(i).Name
      titleSlice = append(titleSlice, columnTitle)
    }

    elementStringValue := rv.Field(i).String()
    tempSlice = append(tempSlice, elementStringValue)
  }

  if !isTitled {
    dataSlice = append(dataSlice, titleSlice)
    isTitled = true
  }
  dataSlice = append(dataSlice, tempSlice)
  }
  ```

3. I'm sorry, what do you mean by non-Incite cases? As for the timestamp I used at `Start`, this is the printout `2022-07-05 13:50:00 +0800 +08`
4. I can't verify for sure, since the CloudWatch Logs doesn't provide the millisecond timestamp
    - Can you help me explain what's the difference between `Chunk` and `SplitUntil`? I've tried reading the documentation but I don't seem to get the difference

**Comment by vcschapp** [comment]
Thanks for all the detail in your reply @lloyd-fftrf. I'll split this response in four parts:

1. Trying to answer some of the questions you raised inline.
2. Explain what I think the problem is.
3. Give a path to a fix and ETA for a fix.

---

### Answers to questions raised inline

Regarding the code snippet you shared (thank you for that, by the way): would you mind sharing the definition of `type logFormat`?

> I'm sorry, what do you mean by non-Incite cases?

By the "non-Incite case", I meant the code you used when you called the CWL Insights API directly, without using Incite. In retrospect, the choice of wording wasn't very clear there.

> I can't verify for sure, since the CloudWatch Logs doesn't provide the millisecond timestamp

This is interesting. The `@timestamp` field, which is a special CWL Insights field representing the timestamp assigned to the log event on submission to CWL, does have millisecond granularity, or it should if the records are submitted to CWL with millisecond granularity.

When I look at your query, I notice you are displaying a custom user field (`time`) but sorting by the CWL Insights log record timestamp (`@timestamp`). It is possible these values differ, and it is possible the `time` field has less granularity than does the `@timestamp` field. To answer the question I asked, you might find it useful to include `@timestamp` in the displayed fields.

```
| display ip, time, http_request, response, url, user_agent, f1, f2 | sort @timestamp asc
              ^^^^                                                         ^^^^^^^^^^
```

> Can you help me explain what's the difference between `Chunk` and `SplitUntil`? I've tried reading the documentation but I don't seem to get the difference

Absolutely! Could I ask in return for your help in critiquing the documentation to see if we can make it more accessible/usable?

The `Chunk` field is used for partitioning the query time range (`Start..End`) into smaller pieces ("chunks") for two reasons:

1. increased parallelism, since Incite can issue many queries to CWL Insights in parallel, one per chunk;
2. improved accuracy, since CWL Insights has a limit of 10K results per query, but Incite converts that into a limit of 10K results per chunk, greatly increasing the number of results you can return from your query.

By itself, the `Chunk` field is "static". So if you specify `Chunk = 5m` and `Start..End = 2022-07-18T09:10:00Z..2022-07-18T09:20:00Z` then Incite will convert your time range into two chunks from 09:10..09:15 and from 09:15..09:20.

The `SplitUntil` field makes the "static" chunk size more dynamic by giving Incite permission to sub-divide your chunks if they are "full", meaning if they receive 10K results. Using the example above, suppose you set `SplitUntil = 30s` and the 09:10..09:15 chunk obtained `Incite.MaxLimit` results (10K). Incite will say to itself "since there were 10K results in that chunk, there might be more than 10K eligible log records, so I will throw away the results and split the chunk into smaller pieces." It will then split the 5m chunk 09:10..09:15 into four chunks of duration 1m15s: 09:10..09:11:15, 09:11:15..09:12:30, and so on, and re-query each sub-divided chunk. If any sub-divided chunk receives 10K results, Incite will keep throwing away the results and splitting it until either the sub-chunks get less than 10K results, or until the sub-chunk duration becomes as small as `SplitUntil`.

Using this adaptive/dynamic chunking mechanism, you can pretty much always configure Incite to get all the results in any arbitrary time range.

Does this make sense, and do you have any suggestions about wording change that would make it clearer within the documentation?

---

### Explanation of what I think the problem is

**tl;dr**: There is a subtle bug in how Incite converts it's chunk ranges to CWL Insights API calls, which is caused by inaccurate documentation in the CloudWatch Logs API reference, and a bug in how the CWL Logs `StartQuery` API works in certain circumstances.

**Details**: The API reference page for `StartQuery` ([link to current version](https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_StartQuery.html#CWL-StartQuery-request-startTime) | [link to Wayback archive from today](https://web.archive.org/web/20220718162458/https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_StartQuery.html)) says that `startTime` and `endTime` are to be specified in *seconds* since the epoch and that the time range is *inclusive* of `endTime`. This is inaccurate:

1. The API actually accepts either seconds *or* milliseconds since the epoch and uses some server-side cleverness to guess which one you supplied.
2. When you specify seconds, the `endTime` is actually only inclusive of the *first millisecond of* the last second.

The Incite code was written following the API reference, so Incite uses seconds-since-the-epoch. This results in Incite "losing" 999 milliseconds worth of search results every chunk. Since your initial query uses a five-minute interval with one-minute chunks, Incite is likely creating five one-minute chunks, so it is probably "losing" 5 x 999ms = 4,995ms = 1.67% of the time range. This estimate is pretty close to the 1.84% of result loss you are seeing IRL (239 / 13,006) so I am pretty confident it is the root of the problem.

I learned about this issue on May 25 and tried to take the easy way out by filing bug reports with the AWS team for CloudWatch to fix both the documentation (#1) and the underlying bug for seconds (#2). I thought they would do this quickly as it is simple, but my issue is still in their input queue.

---

### Path to a fix and ETA for a fix

While IMO this is *technically* a CWL Insights API bug, it can be fixed within Incite and since that can be done far more promptly than waiting for CWL Insights team, I will classify this issue as an Incite bug.

**What needs to be done**: Change the code where Incite calls CWL to convert the chunk time range into milliseconds since epoch instead of seconds since epoch.

**Timeline**: I can make this fix within the next two weeks, or a PR is gratefully accepted if someone else can do it sooner.

---

### Bonus

A bonus of fixing the issue on the Incite side is it will eliminate the limitation that times need to be specified with seconds of granularity. We can increase the granularity down to milliseconds, which will make Incite more flexible and powerful in a backwardly-compatible way.

**Comment by vcschapp** [comment]
Candidate fix commit is pushed (`2faca92`) but it needs testing. Still aiming to complete within the next 12 days.

**Comment by vcschapp** [comment]
This morning I executed the functional testing script below. All tests are passing.

# Functional test setup

We create the following log events...

| Timestamp                | Epoch Millis  | Message                                                  |
|--------------------------|---------------|----------------------------------------------------------|
| 2022-07-22T14:50:00.000Z | 1658501400000 | Begin testing https://github.com/gogama/incite/issues/16 |
| 2022-07-22T14:50:59.001Z | 1658501459001 | Millisecond 1                                            |
| 2022-07-22T14:51:59.500Z | 1658501519500 | Millisecond 500                                          |
| 2022-07-22T14:52:59.999Z | 1658501579999 | Millisecond 999                                          |

...Using the AWS CLI as shown below:

```shell
$ aws --version
aws-cli/2.4.13 Python/3.8.8 Linux/4.4.0-19041-Microsoft exe/x86_64.ubuntu.18 prompt/off

$ aws logs --region us-west-2 create-log-group --log-group-name /github.com/gogama/incite/test

$ aws logs --region us-west-2 create-log-stream \
    --log-group-name /github.com/gogama/incite/test \
    --log-stream-name '2022-07-22@issue#16'

$ aws logs --region us-west-2 describe-log-streams \
    --log-group-name /github.com/gogama/incite/test \
    --log-stream-name '2022-07-22@issue#16' | jq .
{
  "logStreams": [
    {
      "logStreamName": "2022-07-22@issue#16",
      "creationTime": 1658501126264,
      "arn": "arn:aws:logs:us-west-2:XXXXXXXXXXXX:log-group:/github.com/gogama/incite/test:log-stream:2022-07-22@issue#16",
      "storedBytes": 0
    }
  ]
}

$ aws logs --region us-west-2 put-log-events \
    --log-group-name /github.com/gogama/incite/test \
    --log-stream-name '2022-07-22@issue#16' \
    --log-events '[
      {"timestamp":1658501400000,"message":"Begin testing https://github.com/gogama/incite/issues/16"},
      {"timestamp":1658501459001,"message":"Millisecond 1"},
      {"timestamp":1658501519500,"message":"Millisecond 500"},
      {"timestamp":1658501579999,"message":"Millisecond 999"}
    ]' | jq .
{
    "nextSequenceToken": "..............."
}

$ aws logs --region us-west-2 get-log-events\
    --log-group-name /github.com/gogama/incite/test \
    --log-stream-name '2022-07-22@issue#16' \
   | jq -c '.events | .[] | {timestamp, message}'
{"timestamp":1658501400000,"message":"Begin testing https://github.com/gogama/incite/issues/16"}
{"timestamp":1658501459001,"message":"Millisecond 1"}
{"timestamp":1658501519500,"message":"Millisecond 500"}
{"timestamp":1658501579999,"message":"Millisecond 999"}
```

# Functional test execution

We run the following code using Incite, at the following commits:

| Name           | Commit                                                                                                | Start                    | End                      | Chunk | Expected         | Actual           | Status  |
|----------------|-------------------------------------------------------------------------------------------------------|--------------------------|--------------------------|-------|------------------|------------------|---------|
| Before.Bug#1   | [87514ee1](https://github.com/gogama/incite/commit/87514ee1d43328756a76c36b8142995acc494c45) (v1.3.1) | 2022-07-22T14:50:00.000Z | 2022-07-22T14:53:00.000Z | -     | ALL MSG BUT LAST | ALL MSG BUT LAST | ✔ PASS |
| Before.Bug#2   | [87514ee1](https://github.com/gogama/incite/commit/87514ee1d43328756a76c36b8142995acc494c45) (v1.3.1) | 2022-07-22T14:50:00.000Z | 2022-07-22T14:53:00.000Z | 1m    | FIRST MSG ONLY   | FIRST MSG ONLY   | ✔ PASS |
| After.Fix#1    | [2faca92f](https://github.com/gogama/incite/commit/2faca92fd535bacb5542d6b6b670fe2f8e9163a2)          | 2022-07-22T14:50:00.000Z | 2022-07-22T14:53:00.000Z | -     | ALL MSG          | ALL MSG          | ✔ PASS |
| After.Fix#2    | [2faca92f](https://github.com/gogama/incite/commit/2faca92fd535bacb5542d6b6b670fe2f8e9163a2)          | 2022-07-22T14:50:00.000Z | 2022-07-22T14:53:00.000Z | 1m    | ALL MSG          | ALL MSG          | ✔ PASS |
| After.Latest#1 | [ae05c9c9](https://github.com/gogama/incite/commit/ae05c9c9dd0270ab6592544db28dede3e129acea)          | 2022-07-22T14:50:00.000Z | 2022-07-22T14:53:00.000Z | -     | ALL MSG          | ALL MSG          | ✔ PASS |
| After.Latest#2 | [ae05c9c9](https://github.com/gogama/incite/commit/ae05c9c9dd0270ab6592544db28dede3e129acea)          | 2022-07-22T14:50:00.000Z | 2022-07-22T14:53:00.000Z | 1m    | ALL MSG          | ALL MSG          | ✔ PASS |
| After.Latest#3 | [ae05c9c9](https://github.com/gogama/incite/commit/ae05c9c9dd0270ab6592544db28dede3e129acea)          | 2022-07-22T14:50:00.000Z | 2022-07-22T14:53:00.000Z | 250ms | ALL MSG          | ALL MSG          | ✔ PASS |


```go
package main

import (
	"fmt"
	"os"
	"time"

	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/cloudwatchlogs"
	"github.com/gogama/incite"
)

var (
	start               = time.Date(2022, 7, 22, 14, 50, 0, 0, time.UTC)
	end                 = time.Date(2022, 7, 22, 14, 53, 0, 0, time.UTC)
	chunk time.Duration = 0
)

func main() {
	a := cloudwatchlogs.New(session.Must(session.NewSession()))

	m := incite.NewQueryManager(incite.Config{Actions: a})
	defer func() {
		_ = m.Close()
	}()

	s, err := m.Query(incite.QuerySpec{
		Text:   "display @timestamp, @message",
		Start:  start,
		End:    end,
		Groups: []string{"/github.com/gogama/incite/test"},
		Chunk:  chunk,
	})
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	data, err := incite.ReadAll(s)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}

	fmt.Fprintf(os.Stderr, "Got %d results\n", len(data))

	var v []struct {
		Timestamp time.Time `incite:"@timestamp"`
		Message   string    `incite:"@message"`
	}
	err = incite.Unmarshal(data, &v)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}

	fmt.Println(v)
}
```

**Comment by vcschapp** [comment]
Fix will be released with Incite v1.3.2 release, coming sometime in the next week.

In the meantime, it is safe to use the HEAD commit on main.

**Comment by vcschapp** [comment]
@lloyd-fftrf, this issue should now be fixed.

Could you update your code to use the latest release, [v1.3.2](https://github.com/gogama/incite/releases/tag/v1.3.2) and confirm you see all the log events you are expecting?

**Comment by lloyd-fftrf** [comment]
Hi @vcschapp , apologize for taking so long to reply. I've been busy these past few days. Thank you for the update.

I tried running my code again today and I got 13,006 records, which is exactly the same result as manual query on CW Logs Insight. I think the bug is fixed.

> Absolutely! Could I ask in return for your help in critiquing the documentation to see if we can make it more accessible/usable?

Sure, thank you for the explanation. It really helps. For me, I got confused at the first sentence of the `Chunk` documentation. I kind of not sure what `chunked query` meant. I think it would be better if the definition of the `chunked query` is written right after the first sentence. Like below

> // Chunk optionally requests a chunked query and indicates the chunk
> // size.
> 
> // In a chunked query, each chunk is sent to the CloudWatch Logs
> // service as a separate Insights query. This can help large queries
> // complete before the CloudWatch Logs query timeout of 15 minutes,
> // and can increase performance because chunks can be run in parallel.
> 
> // If Chunk is zero, negative, or greater than the difference

This is my opinion, feel free to use it if you think this is appropriate.

**Comment by vcschapp** [comment]
Thanks @lloyd-fftrf! I love the suggested middle paragraph. I created an issue to run a documentation pass and will likely integrate the suggested paragraph more or less verbatim. See [Issue #20](https://github.com/gogama/incite/issues/20).

**Comment by vcschapp** [comment]
Closing this issue as originator has indicated the bug appears to be fixed.
